seastar/src/gate.rs

```diff
@@ -0,0 +1,136 @@
+use cxx::UniquePtr;
+use std::fmt;
+use std::future::Future;
+use std::pin::Pin;
+
+#[cxx::bridge(namespace = "seastar")]
+mod ffi {
+    unsafe extern "C++" {
+        include!("seastar/src/gate.hh");
+
+        type gate;
+        type gate_holder;
+
+        type CloseGateFuture = crate::CloseGateFuture;
+
+        fn new_gate() -> UniquePtr<gate>;
+        fn new_gate_holder(gate: &UniquePtr<gate>) -> Result<UniquePtr<gate_holder>>;
+        fn close_gate(gate: &UniquePtr<gate>) -> CloseGateFuture;
+    }
+}
+
+#[cxx_async::bridge(namespace = seastar)]
+unsafe impl Future for CloseGateFuture {
+    type Output = ();
+}
+
+use ffi::*;
+
+/// Error returned by [try_enter](Gate::try_enter) when called on closed gate.
+pub struct GateClosedError;
+
+impl fmt::Debug for GateClosedError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "GateClosedError: gate closed")
+    }
+}
+
+/// Facility to stop new requests, and to tell when existing requests are done.
+///
+/// When stopping a service that serves asynchronous requests, we are faced with
+/// two problems: preventing new requests from coming in, and knowing when existing
+/// requests have completed. The [Gate] class provides a solution.
+pub struct Gate {
+    gate: UniquePtr<gate>,
+}
+
+/// Facility to hold a gate opened using RAII.
+///
+/// A [GateHolder] is obtained when [try_enter](Gate::try_enter) succeeds.
+///
+/// The [Gate] is left when the [GateHolder] is dropped.
+pub struct GateHolder<'a> {
+    _gate: &'a gate,
+    _holder: UniquePtr<gate_holder>,
+}
+
+impl<'a> Gate {
+    /// Creates a new gate.
+    ///
+    /// The underlying `gate` type has a non-trivial move constructor,
+    /// so we can't return it directly by value.
+    pub fn new() -> Pin<Box<Self>> {
+        Box::pin(Gate { gate: new_gate() })
+    }
+
+    /// Tries to enter the gate.
+    ///
+    /// If it succeeds, it returns [GateHolder] that will leave the gate when destroyed (RAII).
+    ///
+    /// If it fails, it returns [GateClosedError].
+    pub fn try_enter(&'a self) -> Result<GateHolder<'a>, GateClosedError> {
+        GateHolder::new(&self.gate)
+    }
+
+    /// Closes the gate.
+    ///
+    /// It must be called at most once (the underlying implementation aborts if it is closed twice).
+    ///
+    /// Returns a future that resolves after everybody have left the gate.
+    pub async fn close(&self) {
+        close_gate(&self.gate).await.unwrap();
+    }
```

Comment:
After reading this now, I'm wondering: is it possible to create an implementation that leverages Rust's type system to prevent calling close twice?
We could return 2 objects from `Gate::new()` - one would be used to enter the gate, and would be shared, another would be used to close the gate (`GateCloser`?) and would not be shared. That way the close method could take it by value, making it impossible to call it twice.

Reply:
It's definitely possible. I've just tried it and manged to do it by holding the gate in SharedPtr (instead of UniquePtr) by `Gate` and `GateCloser`. This is how it works:
```rust
pub struct Gate {
    gate: SharedPtr<gate>,
}

pub struct GateCloser {
    gate: SharedPtr<gate>,
}

impl<'a> Gate {
    pub fn new() -> (Self, GateCloser) {
        let new_gate = Gate { gate: new_gate() };
        let gate_closer = GateCloser { gate: new_gate.gate.clone() };
        (new_gate, gate_closer)
    }
   
    pub fn try_enter(&'a self) -> Result<GateHolder<'a>, GateClosedError> {
        GateHolder::new(&self.gate)
    }
}

impl GateCloser {
    pub async fn close(self) {
        close_gate(&self.gate).await.unwrap();
    }
}
```
I tried some other implementations:
1. `Gate` still holds `UniquePtr<gate>` and `GateCloser` holds `& UniquePtr<gate>`.
2. `Gate` still holds `UniquePtr<gate>` and `GateCloser` holds `& gate`.

I couldn't make the first one work because I couldn't implement `Gate::new()`. There were compilation errors like "returns a value referencing data owned by the current function". The second one is probably hopeless because calling `seastar::gate::close()` through ffi requires `&mut gate` instead of `& gate`. So the only solution I can think of is `SharedPtr`.

I'm also not sure if preventing multiple calls of `close` like this is really worth it. It complicates our simple interface quite a bit.

Reply:
We discussed this recently with @Lorak-mmk and we don't think we should follow this approach. While the API is slightly nicer, the gate still can be shared with C++ code so this doesn't prevent double close.

Perhaps we could implement a pure-Rust `Gate`; it would make sense for pure Rust applications (i.e. not a component of an existing project). However, in this issue let's concentrate on defining an API for cooperation with a C++ type.

Note: if the gate is closed twice then the second close will abort the process - but it's _still a safe behavior_ according to Rust's definition of safety.